File: 프로그래머스/2/148653. 마법의 엘리베이터/마법의 엘리베이터.java
import java.util.*;

class Solution {
    public int solution(int storey) {
        int answer = 0;
        
        StringBuilder sb = new StringBuilder(String.valueOf(storey));
        sb.reverse();
        
        for (int i = 0; i < sb.length(); i++) {
            int cur = sb.charAt(i) - '0';
            if (cur < 5) {
                answer += cur;
            } else if (cur > 5) {
                answer += (10 - cur);
                if (i != sb.length() - 1) {
                    int n = sb.charAt(i + 1) - '0';
                    sb.setCharAt(i + 1, (char) (n + 1 + '0'));
                } else {
                    answer += 1;
                }
            } else {
                if (i != sb.length() - 1 && sb.charAt(i + 1) >= '5') {
                    int n = sb.charAt(i + 1) - '0';
                    sb.setCharAt(i + 1, (char) (n + 1 + '0'));
                }
                answer += cur;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/148653. 마법의 엘리베이터/마법의 엘리베이터.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <stack>

using namespace std;

int solution(int storey) {
    int answer = 0;
    
    stack<int> st;
    
    string s_num = to_string(storey);
    for(auto& c : s_num)
        st.push(c - '0');
    
    while(!st.empty()) {
        int cur = st.top();
        st.pop();
        
        if(cur < 5) {
            answer += cur;
        } else if(cur > 5) {
            answer += (10 - cur);
            if(!st.empty()) {
                int next = st.top();
                st.pop();
                st.push(next + 1);
            } else {
                st.push(1);
            }
        } else { // cur == 5
            if(!st.empty() && st.top() >= 5) {
                int next = st.top();
                st.pop();
                st.push(next + 1);
            }
            answer += 5;
        }
    }
    
    return answer;
}